### ISSUE TITLE
How to train a torch::jit::script::Module?

### ISSUE THREAD

**Issue by markisus** [pytorch/pytorch#28478]
Existing documentation / tutorials show only how to train a `torch::nn::Module` https://pytorch.org/cppdocs/frontend.html#end-to-end-example

I have attempted to make a training loop in the following manner
```
#include <torch/script.h>
#include <torch/torch.h>
#include <iostream>
#include <vector>
// custom loader code
#include "nets/nets.h"
#include "util/runfiles.h"

int main(int argc, char** argv) {
  std::cout << "Nets example" << std::endl;

  // Custom code that loads the module on CUDA
  auto runfiles = MakeRunfiles(argv[0]);
  torch::jit::script::Module script_module = LoadSegnetBackbone(*runfiles);
  script_module.train();
  std::cout << "Loaded script module" << std::endl;

  // Pull parameters out of the script module so we can push them into the
  // optimizer.
  std::vector<at::Tensor> parameters;
  for (const auto& parameter : script_module.get_parameters()) {
    parameters.push_back(parameter.value().toTensor());
  }
  torch::optim::SGD optimizer(std::move(parameters), /*lr=*/0.01);

  constexpr int kBatchSize = 1;
  for (int epoch = 1; epoch <= 1000; ++epoch) {
    optimizer.zero_grad();

    // The input is a (kBatchSize,3,300,300) tensor filled with ones
    at::Tensor input = torch::ones({kBatchSize, /*channels (rgb) =*/3,
                                    /*height=*/300, /*width=*/300})
                           .to(at::kFloat)
                           .to(at::kCUDA);

    // Push the input through the script module
    std::vector<torch::jit::IValue> inputs;
    inputs.push_back(input);
    at::Tensor script_module_forward = script_module.forward(inputs).toTensor();
    // The result is an output tensor of size (kBatchSize, 32, 300, 300)

    // ground truth is a (kBatchSize, 300, 300) tensor filled with ones
    at::Tensor ground_truth =
        torch::ones({kBatchSize, /*height=*/300, /*width=*/300})
            .to(at::kLong)
            .to(at::kCUDA);

    at::Tensor loss = torch::nll_loss2d(
        torch::log_softmax(script_module_forward, /*dim=*/1), ground_truth);
    loss.backward();
    optimizer.step();

    if (epoch % 50 == 0) {
      std::cout << "Loss was " << loss.item<float>() << std::endl;
    }
  }
}
```

but the loss never changes. I have also posted about this on the pytorch forums. https://discuss.pytorch.org/t/jit-module-parameters-are-not-updating-when-training/58945 

cc @suo @yf225

**Comment by suo** [comment]
@yf225 are the C++ frontend optimizers supposed to work with script modules?

**Comment by yf225** [comment]
@suo Are we able to set the gradients in-place for the parameters obtained from a `torch::jit::script::Module`? If so I think C++ frontend optimizers should work with script modules.

**Comment by lasagnaphil** [comment]
I had a similar issue, and figured out what went wrong.

The method Module::get_parameters() does not return every parameter in the module recursively, it only returns the parameters that the module directly owns. So if module A (with parameter P1) contains module B (with parameter P2), then `A.get_parameters()` returns `{P1}` instead of `{P1, P2}`. (Note that this isn't an issue in the Python frontend, because you can use the option ``recurse=True`` to get all the parameters at once.) To get all the parameters, you have to write the recursive logic yourself:

```
void getModuleParams(std::vector<Tensor>& parameters, const torch::jit::script::Module& module) {
    for (const auto& params : module.get_parameters()) {
        parameters.push_back(params.value().toTensor());
    }
    for (const auto& childModule : module.get_modules()) {
        getModuleParams(parameters, childModule);
    }
}

std::vector<Tensor> parameters;
getModuleParams(parameters, script_module);
```

It would be great if the `recurse` option is implemented in the C++ frontend, to keep the API similar with Python.

**Comment by peiyangL** [comment]
@markisus In libtorch1.4, this issue seems to have been solved. 
```
torch::jit::script::Module module = torch::jit::load(base_path + "vgg11_back.pt");
module.train();
std::vector<at::Tensor> parameters;
for (const auto& params : module.parameters()) {
	parameters.push_back(params);
}
torch::optim::SGD optimizer(parameters, /*lr=*/0.1);
int epoch = 0;
std::vector<torch::jit::IValue> inputs;
inputs.push_back(torch::ones({ 1, 64, 32, 32 }));
torch::Tensor ground_truth = torch::ones({ 1 }).to(at::kLong);
while (epoch < 2) {
	optimizer.zero_grad();
	at::Tensor output = module.forward(inputs).toTensor();
	std::cout << output.slice(/*dim=*/1, /*start=*/0, /*end=*/5) << std::endl;
	torch::Tensor loss = torch::nll_loss(output, ground_truth);
	std::cout << loss << std::endl;
	loss.backward();
	optimizer.step();
	epoch++;
}
```
and the result shows the parameters have been optimized. Like this:
```
0.01 *
 0.5297 -2.4128 -1.3238  7.0693 -1.9776
[ CPUFloatType{1,5} ]
0.0241279
[ CPUFloatType{} ]
-0.0010  0.5087 -0.0157  0.1142 -0.0221
[ CPUFloatType{1,5} ]
-0.508688
[ CPUFloatType{} ]
```

**Comment by wayne391** [comment]
In my case, it worked too!
love PyTorch.

**Comment by jaychen1996-byte** [comment]
Hello,but how should I save my model after training? @wayne391 @peiyangL

**Comment by markisus** [comment]
Haven't tried the new pytorch version myself, but I'll close this out since people are reporting it fixed. Thanks pytorch team!
